lib/adapters/neo4j.js

```diff
@@ -0,0 +1,426 @@
+
+adapter._init = function (options) {
+  // var connectionString = options.connectionString;
+
+  // if (!connectionString || !connectionString.length) {
+  //   connectionString = 'mongodb://' +
+  //     (options.username ? options.username + ':' + options.password + '@' : '') +
+  //     options.host + (options.port ? ':' + options.port : '') + '/' + options.db;
+  // }
+
+  // mongoose.set('debug', options.debug);
+
+  //Setup mongoose instance
+  this.db = new neo4j.GraphDatabase(options.connectionString);
+  
+  console.log("Init DB", this.db);
+};
+
+/**
+ * Store models in an object here.
+ *
+ * @api private
+ */
+adapter._models = {};
+_schema = {};
+
+adapter.schema = function (name, schema, options, schemaCallback) {
+  
+  // console.log("neo4j schema", name, schema, options);
+
+  //TODO : Copied wholemeal from mongoDB - need to think about refactor & re-use
+  options = options || {};
+
+  var refkeys = [];
+  var Mixed = mongoose.Schema.Types.Mixed;
+  var pk = (options.model || {}).pk;
+
+  _.each(schema, function (val, key) {
+    var obj = {};
+    var isArray = _.isArray(val);
+    var value = isArray ? val[0] : val;
+    var isObject = _.isPlainObject(value);
+    var ref = isObject ? value.ref : value;
+    var inverse = isObject ? value.inverse : undefined;
+    var pkType = value.type || value.pkType || mongoose.Schema.Types.ObjectId;
+    var fieldsToIndex = {};
+
+    // Convert strings to associations
+    if (typeof ref === 'string') {
+      var field = _.extend(isObject ? value : {}, {
+        ref: ref,
+        inverse: inverse,
+        type: pkType,
+        external: !!value.external,
+        alias: val.alias || null
+      });
+
+      schema[key] = isArray ? [field] : field;
+
+      refkeys.push(key);
+    }
+
+    // Convert native object to schema type Mixed
+    if (typeof value == 'function' && typeCheck(value) == 'object') {
+
+      if (isObject) {
+        schema[key].type = Mixed;
+      } else {
+        schema[key] = Mixed;
+      }
+    }
+  });
+
+  if(pk){
+    if(_.isFunction(schema[pk])){
+      schema[pk] = { type: schema[pk]};
+    }else if(!(_.isObject(schema[pk]) && schema[pk].type)){
+      throw new Error("Schema PK must either be a type function or an object with a "
+                      + "`type` property");
+    }
+
+    _.extend(schema[pk], {index: {unique: true}});
+  }
+
+  schema = mongoose.Schema(schema, options);
+  schema.refkeys = refkeys;
+
+  _.each(refkeys, function(key){
+    var index = {};
+    index[key] = 1;
+
+    schema.index(index);
+  });
+
+  //Set index on deletedAt
+  schema.index({
+    deletedAt: 1
+  },{
+    sparse: true
+  });
+
+  if (schemaCallback)
+    schemaCallback(schema);
+
+  _schema[name] = schema;
+
+  return schema;
+
+  function typeCheck(fn) {
+    return Object.prototype.toString.call(new fn(''))
+      .slice(1, -1).split(' ')[1].toLowerCase();
+  }
+};
+
+adapter.model = function(name, schema, options) {
```

Comment:
If i remember right, there're some use-cases when this function is used as a getter. It's provided with a string (model name) and expected to return model.